src/checkoutservice/middleware.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	rpcServerDuration syncfloat64.Histogram
+)
+
+func parseFullMethod(fullMethod string) (string, []attribute.KeyValue) {
```

Comment:
Does OpenTelemetry Go provide an instrumentation library for this, or we need to manually write one? What's the best practice? (I think the goal is to show folks best practices, we don't want to show/encourage folks how to write their own HTTP protocol parser).

Reply:
For now, the grpc only has trace instrumentation library in go-contrib.

https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/google.golang.org/grpc/otelgrpc
![image](https://user-images.githubusercontent.com/11855957/186449038-0dddbc67-bb7d-4d6f-9626-1867b447ab3d.png)

So I tried to first add metrics support for grpc service in the demo, and in the meantime I'm trying to add metrics support for grpc in go-contrib repo, too.

What do you think is the best way? We should first add metrics support for grpc into go-contrib and then update demo to use the  instrumentation library of new version? @reyang @austinlparker